src/advanced/timestamping-tutorial.md

```diff
@@ -0,0 +1,929 @@
+the Exonum core.
+
+```rust
+impl From<Error> for ExecutionError {
+    fn from(value: Error) -> ExecutionError {
+        let description = value.to_string();
+        ExecutionError::with_description(value as u8, description)
+    }
+}
+```
+
+The `ExecutionError` will have description the description we have indicated
+for the `Error` enum.
+
+### Transaction Execution
+
+Every transaction in Exonum has business logic of the blockchain attached,
+which is encapsulated in the `Transaction` trait. This trait includes the
+`verify` method to check the integrity of the transaction, and the `execute`
+method which contains logic applied to the storage when a transaction is
+executed. The `Transaction` trait needs to be implemnted for all transactions
+which are to be used in Exonum.
+
+In our case, `verify` for the timestamping transaction checks the transaction
+signature, while `execute` performs the following operations:
+
+1. Gets the current time value provided by the time oracle. If the system
+   cannot get the time value, it will output a corresponding error.
+2. Takes the content of the transaction and calculates its hash.
+3. Checks whether a transaction with the same hash already exists in the
+   blockchain. If a transaction with such a hash is found, the system will
+   output a corresponding error.
+4. Creates a new entry in the database `Fork`, which includes the content of
+   the transaction, its hash and the current time value.
+
+
+```rust
+impl Transaction for TxTimestamp {
+    fn verify(&self) -> bool {
+        self.verify_signature(self.pub_key())
+    }
+
+    fn execute(&self, fork: &mut Fork) -> ExecutionResult {
+        let time = TimeSchema::new(&fork)
+            .time()
+            .get()
+            .expect("Can't get the time");
+
+        let content = self.content();
+        let hash = content.content_hash();
+
+        let mut schema = Schema::new(fork);
+        if let Some(_entry) = schema.timestamps().get(hash) {
+            Err(Error::HashAlreadyExists)?;
+        }
+
+        trace!("Timestamp added: {:?}", self);
+        let entry = TimestampEntry::new(self.content(), &self.hash(), time);
+        schema.add_timestamp(entry);
+        Ok(())
+    }
+}
+```
+
+## Configure API endpoints
+
+Next, we need to implement the node API. We have separated the code that
+configures API endpoints for our Timestamping service into a separate file -
+[src/api.rs][src/api.rs].
+
+First, we need to define the structures required to implement the API endpoints.
+For our Timestamping service, we will define the following three structures:
+
+- the `TimestampQuery` structure which contains only the hash of the
+  timestamping transaction:
+
+  ```rust
+  pub struct TimestampQuery {
+      /// Hash of the requested timestamp.
+      pub hash: Hash,
+  }
+
+  impl TimestampQuery {
+      /// Creates new `TimestampQuery` with given `hash`.
+      pub fn new(hash: Hash) -> Self {
```

Comment:
Side question to @ozkriff: do we really need separate construct given `hash` is public?

Reply:
It's a convention to have these `*Query` structs, as far as I remember.